alicat/serial.py

```diff
@@ -132,21 +132,175 @@ def set_gas(self, gas, retries=2):
+
+        Gas mixes are made using COMPOSER software located from the front panel and over serial.
+        COMPOSER mixes can only be made on Alicat devices with firmware 5v or greater.
+
+        Args:
+            mix_no: The mix number. Gas mixes are stored in slots 236-255. A mix number of 0 will create a mix in
+            the earliest available spot.
+            name: A name for the gas that will appear on the front panel. Names greater than six letters will be
+            cut off.
+            gas#: Name of the gas, as a string. Supported gas types are:
+                'Air', 'Ar', 'CH4', 'CO', 'CO2', 'C2H6', 'H2', 'He', 'N2',
+                'N2O', 'Ne', 'O2', 'C3H8', 'n-C4H10', 'C2H2', 'C2H4',
+                'i-C2H10', 'Kr', 'Xe', 'SF6', 'C-25', 'C-10', 'C-8', 'C-2',
+                'C-75', 'A-75', 'A-25', 'A1025', 'Star29', 'P-5'
+            percent#: The percentage of the mix for the corresponding gas."""
+
+        self._test_controller_open()
+
+        read = '{addr}VE\r'.format(addr=self.address)
+        firmware = self._write_and_read(read, retries)
+        if "2v" in firmware or "3v" in firmware or "4v" in firmware or "GP" in firmware:
+            raise IOError("This unit does not support COMPOSER gas mixes.")
+
+        if mix_no < 236 or mix_no > 255:
+            raise ValueError("Mix number must be between 236-225!")
+
+        total_percent = 0
+        mix_list = [percent1, percent2, percent3, percent4, percent5]
+        for i in range(0, 5):
+            if mix_list[i] is not None:
+                total_percent += mix_list[i]
+        if total_percent != 100:
+            raise ValueError("Percentages of gas mix must add to 100%!")
+
+        if gas1 is not None and gas1 not in self.gases:
+            raise ValueError("{} not supported!".format(gas1))
+        if gas2 is not None and gas2 not in self.gases:
+            raise ValueError("{} not supported!".format(gas2))
+        if gas3 is not None and gas3 not in self.gases:
+            raise ValueError("{} not supported!".format(gas3))
+        if gas4 is not None and gas4 not in self.gases:
+            raise ValueError("{} not supported!".format(gas4))
+        if gas5 is not None and gas5 not in self.gases:
+            raise ValueError("{} not supported!".format(gas5))
+
+        command = '{addr} GM {shortName} {mixNumber} {p1} {g1}' \
+                  ' {p2} {g2} {p3} {g3} {p4} {g4} {p5}' \
+                  ' {g5}\r'.format(addr=self.address, shortName=name, mixNumber=mix_no, p1=percent1,
+                                     g1=self.gases.index(gas1), p2=percent2, g2=self.gases.index(gas2),
+                                     p3=percent3 if gas3 is not None else "", g3=self.gases.index(gas3) if gas3 is not None else "", p4=percent4 if gas4 is not None else "",
+                                     g4=gas4 if gas4 is not None else "", p5=percent5 if gas5 is not None else "", g5=self.gases.index(gas5) if gas5 is not None else "")
```

Comment:
This could also be cleaner and made to not trail so many spaces.

```python
gas_list = ' '.join(' '.join(self.gases.index(gas), percent) for gas, percent in gases.items())
' '.join(self.address, name, mix_no, gas_list) + '\r'
```

Double join read a little awkward but it avoids you repeating yourself. Could replace the inner join with a small format string as well to try helping readability.